File: app/src/main/java/i/am/shiro/amai/service/DownloadService.java
package i.am.shiro.amai.service;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.Nullable;

import com.bumptech.glide.Glide;

import org.apache.commons.io.FileUtils;

import java.io.File;

import i.am.shiro.amai.dao.DownloadQueueManager;
import i.am.shiro.amai.dao.DownloadTaskDispatcher;
import i.am.shiro.amai.model.Book;
import i.am.shiro.amai.model.DownloadTask;
import i.am.shiro.amai.notification.ProgressNotice;
import i.am.shiro.lib.notification.ServiceNotifier;
import timber.log.Timber;

public class DownloadService extends IntentService {

    private final ServiceNotifier notifier = new ServiceNotifier(this, 1);

    public static void start(Context context, Book book) {
        try (DownloadQueueManager queueManager = new DownloadQueueManager()) {
            queueManager.add(book);
        }

        Intent intent = new Intent(context, DownloadService.class);
        context.startService(intent);
    }

    public DownloadService() {
        super(DownloadService.class.getSimpleName());
        setIntentRedelivery(true);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        notifier.startForeground(new ProgressNotice());
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        try (DownloadTaskDispatcher dispatcher = new DownloadTaskDispatcher()) {
            for (DownloadTask task : dispatcher) {
                dispatcher.notifyRunning(task);
                try {
                    runTask(task);
                    dispatcher.notifyDone(task);
                } catch (Exception e) {
                    Timber.w(e);
                    dispatcher.notifyFailed(task);
                }
            }
        }
    }

    private void runTask(DownloadTask task) throws Exception {
        String sourceUrl = task.getSourceUrl();
        File sourceFile = downloadPage(sourceUrl);

        String destinationUrl = task.getDestinationUrl();
        File destinationFile = new File(destinationUrl);

        FileUtils.copyFile(sourceFile, destinationFile);
    }

    private File downloadPage(String pageUrl) throws Exception {
        return Glide.with(this)
                .download(pageUrl)
                .submit()
                .get();
    }
}

File: app/src/main/java/i/am/shiro/amai/service/DownloadService.kt
package i.am.shiro.amai.service

import android.app.Notification
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.os.IBinder
import androidx.core.app.NotificationCompat
import androidx.core.content.getSystemService
import i.am.shiro.amai.AmaiPreferences
import i.am.shiro.amai.DEFAULT_CHANNEL_ID
import i.am.shiro.amai.R
import i.am.shiro.amai.data.AmaiDatabase
import i.am.shiro.amai.data.entity.LocalImageEntity
import i.am.shiro.amai.data.entity.SavedEntity
import okhttp3.OkHttpClient
import okhttp3.Request
import okio.buffer
import okio.sink
import org.koin.android.ext.android.inject
import timber.log.Timber
import java.io.File
import java.io.IOException
import java.util.concurrent.Executors

private const val ID_PROGRESS = 1
private const val ID_DONE = 2
private const val ID_ERROR = 3

class DownloadService : Service() {

    private val database by inject<AmaiDatabase>()

    private val preferences by inject<AmaiPreferences>()

    private val okHttpClient by inject<OkHttpClient>()

    private val notifManager by lazy { getSystemService<NotificationManager>()!! }

    private val executor = Executors.newSingleThreadExecutor()

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int = START_STICKY

    override fun onCreate() {
        super.onCreate()
        startForeground(ID_PROGRESS, buildForegroundNotification())
        executor.execute(::work)
    }

    // TODO resize downloaded image as thumbnail and save it
    private fun work() {
        val download = database.downloadDao.findAvailableJob() ?: return stopSelf()
        val book = database.bookDao.findById(download.bookId)
        val remoteImage = database.remoteImageDao.findByBookIdAndPageIndex(
            download.bookId,
            download.progressIndex
        )

        database.downloadDao.setIsDownloading(download.bookId, true)

        postDownloadProgressNotification(book.title, book.pageCount, download.progressIndex)

        val localFile = File(preferences.storagePath!!)
            .resolve(remoteImage.bookId.toString())
            .apply { mkdirs() }
            .resolve(remoteImage.url.substringAfterLast('/'))

        try {
            downloadPageTo(remoteImage.url, localFile)
            val localImage = LocalImageEntity(
                bookId = download.bookId,
                pageIndex = download.progressIndex,
                width = remoteImage.width,
                height = remoteImage.height,
                url = localFile.path,
                thumbnailWidth = remoteImage.thumbnailWidth,
                thumbnailHeight = remoteImage.thumbnailHeight,
                thumbnailUrl = remoteImage.thumbnailUrl
            )
            database.localImageDao.insert(localImage)
            database.downloadDao.incrementProgress(download.bookId)

            if (download.progressIndex == book.pageCount - 1) {
                database.savedDao.insert(SavedEntity(download.bookId))
                database.downloadDao.setIsDone(download.bookId, true)
                postDownloadDoneNotification(book.title)
            }
        } catch (e: Exception) {
            Timber.e(e)
            database.downloadDao.incrementErrorCount(download.bookId)

            if (download.errorCount + 1 == 3) {
                postDownloadFailedNotification(book.title)
            }
        }

        val nextJob = database.downloadDao.findAvailableJob()
        if (nextJob?.bookId != download.bookId) {
            database.downloadDao.setIsDownloading(download.bookId, false)
        }

        executor.execute(::work)
    }

    @Throws(Exception::class)
    private fun downloadPageTo(pageUrl: String, toFile: File) {
        val request = Request.Builder()
            .url(pageUrl)
            .build()

        okHttpClient.newCall(request).execute().use {
            if (!it.isSuccessful) throw IOException("Unexpected code $it")
            toFile.delete()
            toFile.sink().buffer().writeAll(it.body!!.source())
        }
    }

    private fun buildForegroundNotification(): Notification =
        NotificationCompat.Builder(this, DEFAULT_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notification_download)
            .setContentTitle("Downloading")
            .setProgress(0, 0, true)
            .build()

    private fun postDownloadProgressNotification(title: String, max: Int, progress: Int) {
        NotificationCompat.Builder(this, DEFAULT_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notification_download)
            .setContentTitle("Downloading")
            .setContentText(title)
            .setProgress(max, progress, false)
            .post(ID_PROGRESS)
    }

    private fun postDownloadDoneNotification(title: String) {
        NotificationCompat.Builder(this, DEFAULT_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notification_download)
            .setContentTitle("Download done")
            .setContentText(title)
            .post(ID_DONE)
    }

    private fun postDownloadFailedNotification(title: String) {
        NotificationCompat.Builder(this, DEFAULT_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notification_download)
            .setContentTitle("Download failed")
            .setContentText(title)
            .post(ID_ERROR)
    }

    private fun NotificationCompat.Builder.post(id: Int) = notifManager.notify(id, build())
}

